### ISSUE TITLE
Events from vertx event bus to camel are not processed asynchronously

### ISSUE THREAD

**Issue by tleveque** [vert-x3/vertx-camel-bridge#18]
When a Camel processor execute blocking code, the event loop is blocked.
The root cause is that camel processor process is not async but synchronous even if you use AsyncProcessorConverterHelper to convert the processor to async one as described by Javadoc of the AsyncProcessorConverterHelper class:

A simple converter that can convert any {@link Processor} to an {@link AsyncProcessor}.
Processing will still occur synchronously but it will provide the required
notifications that the caller expects.

The solution would be to ask vertx to execute producer.process call as blocking code into FromVertxToCamelProducer class line 58.
// It's an async producer so won't block.
producer.process(exchange, new CamelProducerCallback(exchange, vertxMessage));

**Comment by cescoffier** [comment]
@davsclaus what do you think ? The issue with this solution is the limitation of the number of concurrent processing. 

In my initial idea I was expecting the route to be non-blocking (using SEDA).

**Comment by davsclaus** [comment]
Yes it depends what happens in Camel as some components are not async by nature. 

If you use seda then you need to set waitForTask to Never to never block.

**Comment by davsclaus** [comment]
Instead of seda we could create a vertx component that is totally non-blocking. Then it also IMHO makes it more clear in this example
https://github.com/vert-x3/vertx-examples/blob/master/camel-bridge-examples/src/main/java/io/vertx/example/camel/feed/FeedExample.java#L47

If it was 

       .addInboundMapping(fromCamel("camel-vertx:announce").toVertx("announce"))
        .addInboundMapping(fromCamel("camel-vertx:errors").toVertx("errors")))

btw Historically seda was designed how it works today, and we can't change it for 2.x. Maybe for Camel 3.0.

**Comment by chrsoo** [comment]
Sorry for asking an existential question in this thread but what is the motivation for adding a `camel-vertx` component using `vertx-camel-bridge` when we already have the [Vertx Component](https://camel.apache.org/vertx.html) in the Camel project?

Technically I guess that the bridge adds a handler directly to the event loop and that the Camel component relies on the event buss, but I fail to see the advantage of supporting integration in both ways. Of the two the Event Bus version is more versatile as it can send messages between Verticles and in the end both ways achieve the same thing.

Maybe I am missing a particular use case that can only be solved one way or the other, but the more natural way to integrate between the two frameworks seems to be the Camel Vert.X Component and not the bridge. With the bridge we are missing the Vert.X and Camel DSL whereas with the Camel component we stay within Camel. The only thing missing is a natural way to start Camel from a Verticle.

Does it not make sense to let the bridge focus only on bootstrapping Camel and communicate only over the Event Bus using the Camel Vertx Component in Camel and using normal Event Bus listeners in Vert.X?

**Comment by cescoffier** [comment]
@davsclaus can give more details about the difference as I don't know much the Vert.x Component in the Camel project.

About the issue, we can add a "isBlocking" flag in the Camel Mapping class so we call the endpoint in an execute blocking construct. It would add this flag and also an optional setter configuring the worker thread pool (using the default Vert.x worker pool by default).

**Comment by davsclaus** [comment]
@chrsoo that is an interesting idea. I guess the existing camel-vertx is for sending/receive from the eventbus and that is likely all we need.

**Comment by cescoffier** [comment]
How would this work for clustered environment, or request-reply ? Does it support that ?

**Comment by davsclaus** [comment]
Ah yeah does the vertx eventbus has some notion of affinity so the reply messages goes back to the same JVM that send the request?

The request/reply code is here in camel-vertx
https://github.com/apache/camel/blob/master/components/camel-vertx/src/main/java/org/apache/camel/component/vertx/VertxProducer.java#L63

**Comment by davsclaus** [comment]
A think to look into is how the camel-vertx component detect/setup the eventbus to use.
https://github.com/apache/camel/blob/master/components/camel-vertx/src/main/java/org/apache/camel/component/vertx/VertxComponent.java

From this bridge we should then configure the component with the existing eventbus or something

**Comment by cescoffier** [comment]
In a request/reply, it goes back to the sender (so yes, same JVM).

**Comment by cescoffier** [comment]
The main difference between the two projects is that this one is an event bus bridge and reuse the concepts Vert.x bridges. It was initially made to let Vert.x apps integrate with other applications using Camel. If you believe that this is not required and can be made in a simpler way with the Vert.x Camel component, let's do this. 

BTW, it looks like the Vert.x Camel component has the same flaw as this one in term of blocking.

**Comment by davsclaus** [comment]
Yes I think the reply handler should not use the thread from the vertx callback but let it execute in its own thread
https://github.com/apache/camel/blob/master/components/camel-vertx/src/main/java/org/apache/camel/component/vertx/VertxProducer.java#L107

Isnt that what you refer to @cescoffier  ?

**Comment by cescoffier** [comment]
it depends on the callback.

If the callback is non-blocking then is must be called on the SAME context as the one having enqueued the request. 

If the callback is blocking, it should use a worker thread and when the result is computed go back to the SAME context as the one having enqueued the request. 

To summarize, the caller must be always executed by the same context (thread).

**Comment by chrsoo** [comment]
To me it seems like the Camel Routes should be executed in a multithreaded Worker Verticle to avoid any problems with blocking IO in Camel components.

It may also be relevant to somehow implement and configure a ThreadPoolFactory, cf [Camel's Threading Model](https://camel.apache.org/threading-model.html).

**Comment by cescoffier** [comment]
No, I don't think so. Multithread worker verticles should never be used as (even supported by Vert.x) they break the threading model of vert.x and lead to a very complicated synchronization protocol. If you want multi-threading, create your own pool but make sure to always call Vert.x code in the right context.

**Comment by davsclaus** [comment]
@chrsoo 

About your comment
https://github.com/vert-x3/vertx-camel-bridge/issues/18#issuecomment-260191060

I have only recently got a bit time to look at Camel and vert.x and building a little example for the Camel in Action book. I did a little football dong simulator that was built first in pure vert.x and then now integrated with Camel as well.

I started with the vertx-camel-bridge as it currently is, but went down to use the camel vertx component. As you say it allows me the Camel routing DSL. But it requires a bit code to setup the vertx instance it should use. 

I lean towards what you are saying to let the vertx-camel-bridge reuse the camel-vertx component and then handle the seamless bootstrapping so the end user don't have to manually setup as much.

My little example is here

pure vertx
https://github.com/camelinaction/camelinaction2/tree/master/chapter7/vertx

vertx and camel
https://github.com/camelinaction/camelinaction2/tree/master/chapter7/vertx-camel

When its more finished and polished I plan to do some blogging about it. The book explains a lot more about the reason for the dong simulator and what me and my friends did in our youth with this kind of game.

Yes the webpage with CSS styling is not my ninja skills.

But the dong audio is the authentic dong we had in danish TVs in the 90's to mid 00's on Saturday football days.

**Comment by cescoffier** [comment]
I've added:

```
OutboundMapping.setBlocking(boolean)
OutboundMapping.setWorkerExecutor(WorkerExecutor)
```

The second method requires the first to be set to `true`. If blocking is set to true, the default worker executor is used.
